base/compiler/utilities.jl

```diff
@@ -104,11 +104,28 @@ function invoke_api(li::CodeInstance)
     return ccall(:jl_invoke_api, Cint, (Any,), li)
 end
 
-function get_staged(li::MethodInstance)
-    may_invoke_generator(li) || return nothing
+function has_opaque_closure(c::CodeInfo)
+    for i = 1:length(c.code)
+        stmt = c.code[i]
+        (isa(stmt, Expr) && stmt.head === :new_opaque_closure) && return true
+    end
+    return false
+end
+
+function get_staged(mi::MethodInstance)
+    may_invoke_generator(mi) || return nothing
+    if isdefined(mi, :uninferred)
+        return copy(mi.uninferred::CodeInfo)
+    end
     try
         # user code might throw errors – ignore them
-        return ccall(:jl_code_for_staged, Any, (Any,), li)::CodeInfo
+        ci = ccall(:jl_code_for_staged, Any, (Any,), mi)::CodeInfo
+        if has_opaque_closure(mi)
+            # For opaque closures, cache the generated code info to make sure
+            # that OpaqueClosure method identity is stable
+            mi.uninferred = copy(ci)
```

Comment:
We already have code to do this in C (jl_copy_ast). Let's keep it there, since this would be unsound otherwise.